src/dmd/dsymbolsem.d

```diff
@@ -6752,3 +6765,154 @@ void aliasSemantic(AliasDeclaration ds, Scope* sc)
+        return;
+    }
+
+    Dsymbol scopesym;
+    Dsymbol as = sc.search(ds.loc, ds.ident, &scopesym);
+    if (!as)
+    {
+        ds.error("undefined identifier `%s`", ds.ident.toChars());
+        return errorRet();
+    }
+    if (as.errors)
+    {
+        return errorRet();
+    }
+
+    AliasDeclaration aliassym = as.isAliasDeclaration();
+    if (!aliassym)
+    {
+        ds.error("identifier `%s` must be an alias declaration", as.toChars());
+        return errorRet();
+    }
+
+    if (aliassym.overnext)
+    {
+        ds.error("cannot reassign overloaded alias");
+        return errorRet();
+    }
+
+    auto aliassymParent = aliassym.toParent();
+    if (aliassymParent != ds.toParent())
+    {
+        ds.error("must have same parent `%s` as alias `%s`", aliassymParent.toChars(), aliassym.toChars());
+        return errorRet();
+    }
+    if (!aliassymParent.isTemplateInstance())
+    {
+        ds.error("must be a member of a template");
+        return errorRet();
+    }
+
+    if (!aliassym.type)
+    {
+        ds.error("alias assignment can only apply to type aliases");
+        return errorRet();
+    }
+
+    if (aliassym.adFlags & Declaration.wasRead)
+    {
+        if (!aliassym.errors)
+            error(ds.loc, "%s was read, so cannot reassign", aliassym.toChars());
+        aliassym.errors = true;
+        return errorRet();
+    }
+
+    const errors = global.errors;
+    aliassym.adFlags |= Declaration.ignoreRead; // temporarilly allow reads of aliassym
+
+    const storage_class = sc.stc & (STC.deprecated_ | STC.ref_ | STC.nothrow_ | STC.nogc | STC.pure_ | STC.shared_ | STC.disable);
+
+//<<>>
+    /* This section is needed because Type.resolve() will:
+     *   const x = 3;
+     *   alias y = x;
+     * try to convert identifier x to 3.
+     */
+    auto s = ds.type.toDsymbol(sc);
+    if (errors != global.errors)
+        return errorRet();
+    if (s == aliassym)
+    {
+        ds.error("cannot resolve");
+        return errorRet();
+    }
+
+    if (!s || !s.isEnumMember())
+    {
+        Type t;
+        Expression e;
+        Scope* sc2 = sc;
+        if (storage_class & (STC.ref_ | STC.nothrow_ | STC.nogc | STC.pure_ | STC.shared_ | STC.disable))
+        {
+            // For 'ref' to be attached to function types, and picked
+            // up by Type.resolve(), it has to go into sc.
+            sc2 = sc.push();
+            sc2.stc |= storage_class & (STC.ref_ | STC.nothrow_ | STC.nogc | STC.pure_ | STC.shared_ | STC.disable);
+        }
+        ds.type = ds.type.addSTC(storage_class);
+        ds.type.resolve(ds.loc, sc2, e, t, s);
+        if (sc2 != sc)
+            sc2.pop();
+
+        if (e)  // Try to convert Expression to Dsymbol
+        {
+            s = getDsymbol(e);
+            if (!s)
+            {
+                if (e.op != TOK.error)
+                    ds.error("cannot alias an expression `%s`", e.toChars());
+                return errorRet();
+            }
+        }
```

Comment:
Can you sync this piece with `AliasSemantic()` for consistency?, a little change I made a month ago.

```
        if (e)  // Try to convert Expression to Dsymbol
        {
            // TupleExp is naturally converted to a TupleDeclaration
            if (auto te = e.isTupleExp())
                s = new TupleDeclaration(te.loc, ds.ident, cast(Objects*)te.exps);
            else
            {
                s = getDsymbol(e);
                if (!s)
                {
                    if (e.op != TOK.error)
                        ds.error("cannot alias an expression `%s`", e.toChars());
                    return errorRet();
                }
            }
        }
```

Reply:
Added to https://github.com/dlang/dmd/pull/12029